src/main/java/org/wildfly/security/auth/server/SecurityIdentity.java

```diff
@@ -410,24 +426,35 @@ public SecurityIdentity createRunAsIdentity(String name) throws AuthorizationExc
         if (name == null) {
             throw log.invalidName();
         }
-        final RunAsPrincipalPermission permission = new RunAsPrincipalPermission(name);
-        if (getPermissions().implies(permission)) {
-            try {
-                final SecurityRealm securityRealm = realmInfo.getSecurityRealm();
-                final RealmIdentity realmIdentity = securityRealm.getRealmIdentity(name, null, null);
-                final AuthorizationIdentity newAuthorizationIdentity = realmIdentity.getAuthorizationIdentity();
-                SecurityRealm.safeHandleRealmEvent(securityRealm, new RealmIdentitySuccessfulAuthorizationEvent(this.authorizationIdentity, this.principal, principal));
+        if (authorize) {
+            final RunAsPrincipalPermission permission = new RunAsPrincipalPermission(name);
+            if (! getPermissions().implies(permission)) {
+                SecurityRealm.safeHandleRealmEvent(realmInfo.getSecurityRealm(), new RealmIdentityFailedAuthorizationEvent(authorizationIdentity, this.principal, principal));
+                throw log.unauthorizedRunAs(this.principal, principal, permission);
+            }
+        } else {
+            final SecurityManager sm = System.getSecurityManager();
+            if (sm != null) {
                 try {
-                    return new SecurityIdentity(domain, principal, realmInfo, newAuthorizationIdentity, roleMappers);
-                } finally {
-                    realmIdentity.dispose();
+                    sm.checkPermission(SET_RUN_AS_PERMISSION);
+                } catch (SecurityException ex) {
+                    SecurityRealm.safeHandleRealmEvent(realmInfo.getSecurityRealm(), new RealmIdentityFailedAuthorizationEvent(authorizationIdentity, this.principal, principal));
+                    throw log.unauthorizedRunAs(this.principal, principal, SET_RUN_AS_PERMISSION);
```

Comment:
I think this exception should only be used for user authorization failure, not code authorization failure (which was covered by the existing AccessControlException which is already produced).